From 7cdeef22fb3164bbb52d9fcec9186c6eda877eef Mon Sep 17 00:00:00 2001
From: Dave Taht <dave.taht@bufferbloat.net>
Date: Mon, 27 Aug 2012 10:55:45 -0700
Subject: [PATCH 2/4] net: add skb_reduce_truesize

skb_reduce_truesize is a helper function for shrinking skbs whenever
needed.
---
 include/linux/skbuff.h |   21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/include/linux/skbuff.h b/include/linux/skbuff.h
index 42854ce..38f7bb8 100644
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -2531,5 +2531,26 @@ static inline bool skb_is_recycleable(const struct sk_buff *skb, int skb_size)
 
 	return true;
 }
+
+/*
+ * Caller wants to reduce memory needs before queueing skb
+ * The (expensive) copy should not be be done in fast path.
+ */
+
+static inline struct sk_buff *skb_reduce_truesize(struct sk_buff *skb)
+{
+       if (skb->truesize > 2 * SKB_TRUESIZE(skb->len)) {
+               struct sk_buff *nskb;
+
+               nskb = skb_copy_expand(skb, skb_headroom(skb), 0,
+                                      GFP_ATOMIC | __GFP_NOWARN);
+               if (nskb) {
+                       __kfree_skb(skb);
+                       skb = nskb;
+               }
+       }
+       return skb;
+}
+
 #endif	/* __KERNEL__ */
 #endif	/* _LINUX_SKBUFF_H */
-- 
1.7.9.5

